src/ng/browser.js

```diff
@@ -143,7 +143,7 @@ function Browser(window, document, $log, $sniffer, $$taskTrackerFactory) {
       // - pendingLocation is needed as browsers don't allow to read out
       //   the new location.href if a reload happened or if there is a bug like in iOS 9 (see
       //   https://openradar.appspot.com/22186109).
-      return pendingLocation || location.href;
+      return pendingLocation || trimEmptyHash(location.href);
```

Comment:
What about `pendingLocation`?

Reply:
`pendingLocation` always has the empty hash fragment stripped. In a nutshell, `pendingLocation` is always the value `$browser.url(url)` is called with, which is always `$location.absUrl()` (directly or indirectly), which always has the empty hash stripped off.

This potential difference between `pendingLocation` and `location.href` is what causes the issue. Because while `pendingLocation` is defined, `$browser.url()` return the url without the empty hash, but once `pendingLocation` gets undefined (i.e. on `hashchange`, `popstate`, which may be triggered asynchronously), then `$browser.url()` starts returning the url with the empty fragment.

So, the returned url changes, even if there is no actual change in the url.

Reply:
> This potential difference between pendingLocation and location.href is what causes the issue

Aha, that helps me understand this a bit more. Why does `location.href` have the hash in the first place?

Reply:
That's what browsers do (sometimes :grin:). I found that this can happen when clicking a link with a trailing # (but not when directly setting `location.hash`).

Reply:
I think I'd vote to still do `trimEmptyHash(pendingLocation || location.href)` just for consistency...

Reply:
Done.